WireDomain/Sources/WireDomain/Repositories/FeatureConfigRepository.swift

```diff
@@ -0,0 +1,363 @@
+
+import Combine
+import WireAPI
+import WireDataModel
+
+/// Facilitates access to feature configs related domain objects.
+
+protocol FeatureConfigRepositoryProtocol {
+
+    /// Pulls feature configs from the server and stores them locally.
+    ///
+
+    func pullFeatureConfigs() async throws
+
+    /// Observes feature states.
+    ///
+    /// Each time `pullFeatureConfigs()` is called, a feature config is
+    /// stored locally and a new `FeatureState` value is produced by the publisher.
+    /// It allows the user to be notified of any feature changes over time.
+    ///
+    /// - Warning:  Use this method before calling `pullFeatureConfigs` to receive all emitted values.
+    ///
+    /// - Returns: A publisher of `FeatureState`.
+
+    func observeFeatureStates() -> AnyPublisher<FeatureState, Never>
+
+    /// Fetches a feature config locally.
+    ///
+    /// - Parameter name: The feature name to fetch the config for.
+    /// - Parameter type: The type of config to retrieve.
+    /// - Returns: A `LocalFeature` object with a status and a config (if any).
+
+    func fetchFeatureConfig<T: Decodable>(withName name: Feature.Name, type: T.Type) async throws -> LocalFeature<T>
+
+    /// Fetches a flag indicating whether the user should be notified of a given feature.
+    /// - Parameter name: The feature name.
+    /// - Returns: `true` if user should be notified.
+
+    func fetchNeedsToNotifyUser(forFeatureName name: Feature.Name) async throws -> Bool
+
+    /// Stores a flag indicating whether the user should be notified of a given feature.
+    /// - Parameter notifyUser: Whether the user should be notified for a given feature.
+    /// - Parameter name: The name of the feature to set the flag for.
+
+    func storeNeedsToNotifyUser(_ notifyUser: Bool, forFeatureName name: Feature.Name) async throws
+
+}
+
+final class FeatureConfigRepository: FeatureConfigRepositoryProtocol {
+
+    // MARK: - Properties
+
+    private let featureConfigsAPI: any FeatureConfigsAPI
+    private let context: NSManagedObjectContext
+    private let featureStateSubject = PassthroughSubject<FeatureState, Never>()
+
+    // MARK: - Object lifecycle
+
+    init(
+        featureConfigsAPI: any FeatureConfigsAPI,
+        context: NSManagedObjectContext
+    ) {
+        self.featureConfigsAPI = featureConfigsAPI
+        self.context = context
+    }
+
+    // MARK: - Public
+
+    func pullFeatureConfigs() async throws {
+        let featureConfigs = try await featureConfigsAPI.getFeatureConfigs()
+
+        for featureConfig in featureConfigs {
+            do {
+                try await storeFeatureConfig(featureConfig)
+
+                if let featureState = try await getFeatureState(
+                    forFeatureConfig: featureConfig
+                ) {
+                    featureStateSubject.send(featureState)
+                }
+            } catch {
+                throw FeatureConfigRepositoryError.failedToStoreConfigLocally(error)
```

Comment:
so if we fail to store one config we abort on the others, is it intended?

Reply:
The error is thrown when a a config could not be encoded which should be an edge case since the passed parameter is of type any Codable. In any cases, you're right that this might be not ideal to abort the others if that happens. I will provide something else to avoid that, thanks 👍

Reply:
fixed